File: adapter/py/4_legacy_system.py
"""
Adapter Pattern - 舊系統資料遷移範例

情境：
- 公司從舊系統升級到新系統
- 舊系統使用 XML 格式儲存資料
- 新系統使用 JSON 格式
- 需要在不修改舊系統的情況下遷移資料
"""

from abc import ABC, abstractmethod
from typing import List, Dict, Any
from datetime import datetime
import json


# ============================================================
# 第一部分：舊系統 (XML) - 不可修改
# ============================================================

class LegacyXMLReader:
    """舊系統的 XML 資料讀取器 (不可修改)"""

    def __init__(self):
        # 模擬從資料庫或檔案讀取的 XML 資料
        self._xml_data = """
        <?xml version="1.0" encoding="UTF-8"?>
        <orders>
            <order>
                <order_no>ORD-2024-001</order_no>
                <order_date>2024-01-15</order_date>
                <customer>
                    <cust_id>C001</cust_id>
                    <name>王小明</name>
                    <email>wang@example.com</email>
                    <phone>0912-345-678</phone>
                </customer>
                <items>
                    <item>
                        <product_code>P001</product_code>
                        <product_name>iPhone 15</product_name>
                        <quantity>1</quantity>
                        <unit_price>35000</unit_price>
                    </item>
                    <item>
                        <product_code>P002</product_code>
                        <product_name>AirPods Pro</product_name>
                        <quantity>2</quantity>
                        <unit_price>7500</unit_price>
                    </item>
                </items>
                <total_amount>50000</total_amount>
                <status>completed</status>
            </order>
            <order>
                <order_no>ORD-2024-002</order_no>
                <order_date>2024-01-16</order_date>
                <customer>
                    <cust_id>C002</cust_id>
                    <name>李小華</name>
                    <email>lee@example.com</email>
                    <phone>0988-888-888</phone>
                </customer>
                <items>
                    <item>
                        <product_code>P003</product_code>
                        <product_name>MacBook Pro</product_name>
                        <quantity>1</quantity>
                        <unit_price>65000</unit_price>
                    </item>
                </items>
                <total_amount>65000</total_amount>
                <status>pending</status>
            </order>
        </orders>
        """

    def get_orders_as_xml(self) -> str:
        """回傳 XML 格式的訂單資料"""
        return self._xml_data

    def parse_xml(self, xml_string: str) -> List[Dict]:
        """解析 XML (實際應用會用 xml.etree.ElementTree 或 lxml)"""
        # 簡化的 XML 解析 (實際會更複雜)
        orders = []

        # 模擬解析過程
        # 實際專案會使用: import xml.etree.ElementTree as ET
        # tree = ET.fromstring(xml_string)

        # 這裡直接回傳模擬的字典格式
        return [
            {
                "order_no": "ORD-2024-001",
                "order_date": "2024-01-15",
                "customer": {
                    "cust_id": "C001",
                    "name": "王小明",
                    "email": "wang@example.com",
                    "phone": "0912-345-678"
                },
                "items": [
                    {
                        "product_code": "P001",
                        "product_name": "iPhone 15",
                        "quantity": 1,
                        "unit_price": 35000
                    },
                    {
                        "product_code": "P002",
                        "product_name": "AirPods Pro",
                        "quantity": 2,
                        "unit_price": 7500
                    }
                ],
                "total_amount": 50000,
                "status": "completed"
            },
            {
                "order_no": "ORD-2024-002",
                "order_date": "2024-01-16",
                "customer": {
                    "cust_id": "C002",
                    "name": "李小華",
                    "email": "lee@example.com",
                    "phone": "0988-888-888"
                },
                "items": [
                    {
                        "product_code": "P003",
                        "product_name": "MacBook Pro",
                        "quantity": 1,
                        "unit_price": 65000
                    }
                ],
                "total_amount": 65000,
                "status": "pending"
            }
        ]


# ============================================================
# 第二部分：新系統介面 (Target Interface)
# ============================================================

class DataConverter(ABC):
    """資料轉換介面 - 新系統期望的格式"""

    @abstractmethod
    def convert_orders(self) -> List[Dict]:
        """
        轉換訂單資料
        Returns:
            新系統期望的 JSON 格式:
            [
                {
                    "orderId": "string",
                    "createdAt": "ISO8601 datetime",
                    "customer": {
                        "customerId": "string",
                        "name": "string",
                        "contact": {
                            "email": "string",
                            "phone": "string"
                        }
                    },
                    "lineItems": [
                        {
                            "productId": "string",
                            "name": "string",
                            "quantity": "number",
                            "price": "number"
                        }
                    ],
                    "totalPrice": "number",
                    "orderStatus": "string (pending|processing|completed|cancelled)"
                }
            ]
        """
        pass

    @abstractmethod
    def export_to_json(self, filepath: str) -> bool:
        """匯出為 JSON 檔案"""
        pass


# ============================================================
# 第三部分：轉接器實作
# ============================================================

class LegacyToJsonAdapter(DataConverter):
    """舊系統 XML 轉新系統 JSON 的轉接器"""

    # 狀態映射表 (舊系統 → 新系統)
    STATUS_MAP = {
        "completed": "completed",
        "pending": "pending",
        "cancelled": "cancelled",
        "processing": "processing",
        "shipped": "processing"
    }

    def __init__(self, legacy_reader: LegacyXMLReader):
        self._legacy = legacy_reader
        self._converted_data: List[Dict] = []

    def convert_orders(self) -> List[Dict]:
        """執行轉換邏輯"""

        # 1. 讀取舊系統 XML 資料
        raw_orders = self._legacy.parse_xml(
            self._legacy.get_orders_as_xml()
        )

        # 2. 轉換為新系統格式
        self._converted_data = []
        for order in raw_orders:
            converted_order = self._convert_single_order(order)
            self._converted_data.append(converted_order)

        return self._converted_data

    def _convert_single_order(self, old_order: Dict) -> Dict:
        """轉換單筆訂單"""

        # 轉換訂單編號格式
        # 舊: ORD-2024-001 → 新: ORD-2024-0001
        order_id = self._format_order_id(old_order["order_no"])

        # 轉換日期格式
        # 舊: 2024-01-15 → 新: 2024-01-15T00:00:00Z
        created_at = self._format_datetime(old_order["order_date"])

        # 轉換客戶資料結構
        customer = {
            "customerId": old_order["customer"]["cust_id"],
            "name": old_order["customer"]["name"],
            "contact": {
                "email": old_order["customer"]["email"],
                "phone": old_order["customer"]["phone"]
            }
        }

        # 轉換商品資料結構
        line_items = []
        for item in old_order["items"]:
            line_items.append({
                "productId": item["product_code"],
                "name": item["product_name"],
                "quantity": item["quantity"],
                "price": item["unit_price"]
            })

        # 轉換訂單狀態
        old_status = old_order["status"].lower()
        order_status = self.STATUS_MAP.get(old_status, "pending")

        return {
            "orderId": order_id,
            "createdAt": created_at,
            "customer": customer,
            "lineItems": line_items,
            "totalPrice": old_order["total_amount"],
            "orderStatus": order_status
        }

    def _format_order_id(self, old_id: str) -> str:
        """格式化訂單編號"""
        # ORD-2024-001 → ORD-2024-0001
        parts = old_id.split("-")
        if len(parts) == 3:
            return f"{parts[0]}-{parts[1]}-{int(parts[2]):04d}"
        return old_id

    def _format_datetime(self, date_str: str) -> str:
        """格式化日期時間為 ISO8601 格式"""
        # 假設輸入是 "2024-01-15"
        return f"{date_str}T00:00:00Z"

    def export_to_json(self, filepath: str) -> bool:
        """匯出為 JSON 檔案"""
        if not self._converted_data:
            self.convert_orders()

        try:
            with open(filepath, 'w', encoding='utf-8') as f:
                json.dump(self._converted_data, f, ensure_ascii=False, indent=2)
            return True
        except Exception as e:
            print(f"匯出失敗: {e}")
            return False


# ============================================================
# 第四部分：執行範例
# ============================================================

def display_comparison(old_data: List[Dict], new_data: List[Dict]) -> None:
    """顯示轉換前後的比較"""

    print("\n" + "=" * 80)
    print("📊 資料轉換比較")
    print("=" * 80)

    print("\n【舊系統格式 (XML 解析後的字典)】")
    print("-" * 40)
    for order in old_data[:1]:  # 只顯示第一筆
        print(f"訂單編號: {order['order_no']}")
        print(f"日期: {order['order_date']}")
        print(f"客戶: {order['customer']['name']}")
        print(f"商品: {len(order['items'])} 項")
        print(f"總金額: {order['total_amount']}")
        print(f"狀態: {order['status']}")

    print("\n【新系統格式 (JSON)】")
    print("-" * 40)
    if new_data:
        order = new_data[0]
        print(f"orderId: {order['orderId']}")
        print(f"createdAt: {order['createdAt']}")
        print(f"customer: {order['customer']}")
        print(f"lineItems: {len(order['lineItems'])} 項")
        print(f"totalPrice: {order['totalPrice']}")
        print(f"orderStatus: {order['orderStatus']}")

    print()


def display_json_output(data: List[Dict]) -> None:
    """顯示 JSON 輸出格式"""

    print("\n【JSON 輸出範例】")
    print("-" * 40)
    print(json.dumps(data[0], ensure_ascii=False, indent=2))


if __name__ == "__main__":
    print("=" * 80)
    print("Adapter Pattern - 舊系統資料遷移範例")
    print("=" * 80)

    # 建立舊系統讀取器 (不可修改)
    legacy_reader = LegacyXMLReader()

    # 建立轉接器
    adapter = LegacyToJsonAdapter(legacy_reader)

    # 執行轉換
    old_data = legacy_reader.parse_xml(legacy_reader.get_orders_as_xml())
    new_data = adapter.convert_orders()

    # 顯示轉換結果
    display_comparison(old_data, new_data)
    display_json_output(new_data)

    # 匯出 JSON
    success = adapter.export_to_json("converted_orders.json")
    if success:
        print("\n✅ 已成功匯出至 converted_orders.json")

    print("\n🎯 重點說明：")
    print("1. 舊系統完全不需要修改")
    print("2. Adapter 負責處理格式轉換")
    print("3. 新系統只需要知道 DataConverter 介面")
    print("4. 未來如有其他舊格式，只需新增對應的 Adapter")
    print("=" * 80)

File: adapter/cs/4_LegacySystem.cs
namespace AdapterExamples
{
    // ============================================================
    // 第四部分：舊系統資料遷移範例
    // ============================================================

    #region 舊系統 (XML) - 不可修改

    /// <summary>
    /// 舊系統的 XML 資料讀取器 (不可修改)
    /// </summary>
    public class LegacyXmlReader
    {
        /// <summary>
        /// 模擬從資料庫讀取的 XML 資料
        /// </summary>
        public List<Dictionary<string, object>> GetOrdersAsDictionary()
        {
            // 模擬解析後的資料 (實際會用 XML 解析器)
            return new List<Dictionary<string, object>>
            {
                new Dictionary<string, object>
                {
                    ["order_no"] = "ORD-2024-001",
                    ["order_date"] = "2024-01-15",
                    ["customer"] = new Dictionary<string, string>
                    {
                        ["cust_id"] = "C001",
                        ["name"] = "王小明",
                        ["email"] = "wang@example.com",
                        ["phone"] = "0912-345-678"
                    },
                    ["items"] = new List<Dictionary<string, object>>
                    {
                        new Dictionary<string, object>
                        {
                            ["product_code"] = "P001",
                            ["product_name"] = "iPhone 15",
                            ["quantity"] = 1,
                            ["unit_price"] = 35000
                        },
                        new Dictionary<string, object>
                        {
                            ["product_code"] = "P002",
                            ["product_name"] = "AirPods Pro",
                            ["quantity"] = 2,
                            ["unit_price"] = 7500
                        }
                    },
                    ["total_amount"] = 50000,
                    ["status"] = "completed"
                },
                new Dictionary<string, object>
                {
                    ["order_no"] = "ORD-2024-002",
                    ["order_date"] = "2024-01-16",
                    ["customer"] = new Dictionary<string, string>
                    {
                        ["cust_id"] = "C002",
                        ["name"] = "李小華",
                        ["email"] = "lee@example.com",
                        ["phone"] = "0988-888-888"
                    },
                    ["items"] = new List<Dictionary<string, object>>
                    {
                        new Dictionary<string, object>
                        {
                            ["product_code"] = "P003",
                            ["product_name"] = "MacBook Pro",
                            ["quantity"] = 1,
                            ["unit_price"] = 65000
                        }
                    },
                    ["total_amount"] = 65000,
                    ["status"] = "pending"
                }
            };
        }
    }

    #endregion

    #region 新系統介面 (Target Interface)

    /// <summary>
    /// 資料轉換介面 - 新系統期望的格式
    /// </summary>
    public interface IDataConverter
    {
        /// <summary>
        /// 轉換訂單資料
        /// </summary>
        List<OrderDto> ConvertOrders();

        /// <summary>
        /// 匯出為 JSON 檔案
        /// </summary>
        bool ExportToJson(string filepath);
    }

    /// <summary>
    /// 新系統訂單資料結構
    /// </summary>
    public record OrderDto(
        string OrderId,
        string CreatedAt,
        CustomerDto Customer,
        List<LineItemDto> LineItems,
        decimal TotalPrice,
        string OrderStatus
    );

    public record CustomerDto(
        string CustomerId,
        string Name,
        ContactDto Contact
    );

    public record ContactDto(
        string Email,
        string Phone
    );

    public record LineItemDto(
        string ProductId,
        string Name,
        int Quantity,
        decimal Price
    );

    #endregion

    #region 轉接器實作

    /// <summary>
    /// 舊系統 XML 轉新系統 JSON 的轉接器
    /// </summary>
    public class LegacyToJsonAdapter : IDataConverter
    {
        private readonly LegacyXmlReader _legacyReader;
        private List<OrderDto> _convertedData = new();

        // 狀態映射表
        private static readonly Dictionary<string, string> StatusMap = new()
        {
            { "completed", "completed" },
            { "pending", "pending" },
            { "cancelled", "cancelled" },
            { "processing", "processing" },
            { "shipped", "processing" }
        };

        public LegacyToJsonAdapter(LegacyXmlReader legacyReader)
        {
            _legacyReader = legacyReader;
        }

        public List<OrderDto> ConvertOrders()
        {
            // 1. 讀取舊系統資料
            var rawOrders = _legacyReader.GetOrdersAsDictionary();

            // 2. 轉換為新系統格式
            _convertedData = rawOrders.Select(ConvertSingleOrder).ToList();

            return _convertedData;
        }

        private OrderDto ConvertSingleOrder(Dictionary<string, object> oldOrder)
        {
            // 轉換訂單編號格式
            string orderId = FormatOrderId(oldOrder["order_no"].ToString()!);

            // 轉換日期格式
            string createdAt = FormatDateTime(oldOrder["order_date"].ToString()!);

            // 轉換客戶資料結構
            var oldCustomer = (Dictionary<string, string>)oldOrder["customer"];
            var customer = new CustomerDto(
                CustomerId: oldCustomer["cust_id"],
                Name: oldCustomer["name"],
                Contact: new ContactDto(
                    Email: oldCustomer["email"],
                    Phone: oldCustomer["phone"]
                )
            );

            // 轉換商品資料結構
            var oldItems = (List<Dictionary<string, object>>)oldOrder["items"];
            var lineItems = oldItems.Select(item => new LineItemDto(
                ProductId: item["product_code"].ToString()!,
                Name: item["product_name"].ToString()!,
                Quantity: Convert.ToInt32(item["quantity"]),
                Price: Convert.ToDecimal(item["unit_price"])
            )).ToList();

            // 轉換訂單狀態
            string oldStatus = oldOrder["status"].ToString()!.ToLower();
            string orderStatus = StatusMap.GetValueOrDefault(oldStatus, "pending");

            return new OrderDto(
                OrderId: orderId,
                CreatedAt: createdAt,
                Customer: customer,
                LineItems: lineItems,
                TotalPrice: Convert.ToDecimal(oldOrder["total_amount"]),
                OrderStatus: orderStatus
            );
        }

        private string FormatOrderId(string oldId)
        {
            // ORD-2024-001 → ORD-2024-0001
            var parts = oldId.Split('-');
            if (parts.Length == 3 && int.TryParse(parts[2], out int num))
            {
                return $"{parts[0]}-{parts[1]}-{num:D4}";
            }
            return oldId;
        }

        private string FormatDateTime(string dateStr)
        {
            // 2024-01-15 → 2024-01-15T00:00:00Z
            return $"{dateStr}T00:00:00Z";
        }

        public bool ExportToJson(string filepath)
        {
            try
            {
                if (_convertedData.Count == 0)
                {
                    ConvertOrders();
                }

                var json = System.Text.Json.JsonSerializer.Serialize(
                    _convertedData,
                    new System.Text.Json.JsonSerializerOptions
                    {
                        WriteIndented = true,
                        Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping
                    }
                );

                File.WriteAllText(filepath, json);
                return true;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"匯出失敗: {ex.Message}");
                return false;
            }
        }
    }

    #endregion

    /// <summary>
    /// 舊系統遷移範例執行
    /// </summary>
    public static class LegacySystem
    {
        public static void Run()
        {
            Console.WriteLine("\n" + "=".PadRight(60, '='));
            Console.WriteLine("4. 舊系統資料遷移範例");
            Console.WriteLine("=".PadRight(60, '='));

            // 建立舊系統讀取器
            var legacyReader = new LegacyXmlReader();

            // 建立轉接器
            var adapter = new LegacyToJsonAdapter(legacyReader);

            // 執行轉換
            var newData = adapter.ConvertOrders();

            // 顯示第一筆資料的轉換結果
            var first = newData[0];

            Console.WriteLine("\n【舊系統格式】");
            Console.WriteLine($"  訂單編號: ORD-2024-001");
            Console.WriteLine($"  日期: 2024-01-15");
            Console.WriteLine($"  客戶: 王小明");
            Console.WriteLine($"  商品: {first.LineItems.Count} 項");
            Console.WriteLine($"  總金額: {first.TotalPrice:N0}");
            Console.WriteLine($"  狀態: completed");

            Console.WriteLine("\n【新系統格式 (JSON)】");
            Console.WriteLine($"  orderId: {first.OrderId}");
            Console.WriteLine($"  createdAt: {first.CreatedAt}");
            Console.WriteLine($"  customer: {first.Customer.Name}");
            Console.WriteLine($"  lineItems: {first.LineItems.Count} 項");
            Console.WriteLine($"  totalPrice: {first.TotalPrice}");
            Console.WriteLine($"  orderStatus: {first.OrderStatus}");

            Console.WriteLine("\n【JSON 範例輸出】");
            Console.WriteLine(System.Text.Json.JsonSerializer.Serialize(first,
                new System.Text.Json.JsonSerializerOptions { WriteIndented = true }));

            // 匯出 JSON
            bool success = adapter.ExportToJson("converted_orders.json");
            if (success)
            {
                Console.WriteLine("\n✅ 已成功匯出至 converted_orders.json");
            }

            Console.WriteLine("\n🎯 重點說明：");
            Console.WriteLine("   1. 舊系統完全不需要修改");
            Console.WriteLine("   2. Adapter 負責處理格式轉換");
            Console.WriteLine("   3. 新系統只需要知道 IDataConverter 介面");
            Console.WriteLine("   4. 未來如有其他舊格式，只需新增對應的 Adapter");
        }
    }
}